active/0000-build-fastboot-in-ember-cli.md

```diff
@@ -0,0 +1,75 @@
+
+In order for your app/addon to be able to run in a Node environment, it needs to be compatible in such environment. Compatibility includes not doing DOM/window access in Node, making sure you are able to make API requests in Node etc. Therefore an app/addon may need to provide fastboot specific initializers, instance-initializers, services etc that are only applicable in the Node environment and not in the browser. Including them in the browser has performance implication since you will send down more bytes to the client.
+
+Currently fastboot builds the app twice and generates a different set of assets for fastboot and browser environments via ember fastboot command. Most of the app between the two sets of assets is same except where we want fastboot to override the browser behavior or need fastboot specific handling. Currently ember fastboot filters these and generates the two sets of assets. Building different sets of assets (app and vendor files) for different environments cannot guarantee the app works correctly in both environments always. Moreover, ember-fastboot --serve-assets command also spins up a node server and serves your assets. This is very similar to what ember serve does today. However, we end up forking and creating two parallel worlds with problems such as serving stale assets, creating race conditions during serving and building etc. We want to be able converge the two worlds such that fastboot should be able to specify the last middleware that will be used to serve the assets in fastboot. This helps us not create fastboot commands and keeps the eco-system unified.
+
+We want to build these fastboot specific code as an additive asset to the browser that loads in the Node VM in addtion to the the other assets. If it overrides any browser behavior, having the same module id will override the definition in the loader. In order to build fastboot assets in a performant way, we would like to integrate building fastboot into ember-cli. This issue specifically tracks integrating building fastboot in ember-cli. When we need to build more such artifacts, we can extend this to be a generic target solution.
+
+# Detailed design
+
+## Directory Structure for fastboot app/addons
+In order to build the fastboot assets which are an addition to the app and vendor files, we want to define a separate directory structure. This will allow fastboot specific app code to live in its own namespace and does not touch the browser specific app file.
+
+Currently today no fastboot specific code lives in the addon namespace. Hence we do not need `fastboot/addon` namespace as part of this RFC in `ember-cli`.
```

Comment:
@rwjblue If I understand correctly today:
1. No fastboot specific code lives in addon name space (including `ember-cli-head`
2. `fastboot-filter-initializers` broccoli plugin only filter out instance-initializers and initializers and I don't see why someone would have those in addon namespace.

In general the goal of this RFC was:
1. To unblock Fastboot 1.0
2. Do the minimal work in ember-cli such that fastboot can be built. 

While I agree we can have fastboot specific things in addon namespace (there is nothing prevents that today), there is no usecase I have seen that is happening currently. I really want to address what is needed to unblock Fastboot 1.0. Having, addon namespace for fastboot is *good to have* feature but not a *required* feature.

I can add a long term section of allowing building `fastboot/addon` if you want.